gds_write/build_defs.bzl

```diff
@@ -0,0 +1,74 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""These build rules run automated GDS write on an implemented design"""
+
+load("//place_and_route:open_road.bzl", "OpenRoadInfo")
+load("//synthesis:build_defs.bzl", "SynthesisInfo")
+
+def _gds_write_impl(ctx):
+    final_gds = ctx.actions.declare_file("{}.gds".format(ctx.attr.name))
+
+    lef_args = " "
+    for file in ctx.files.input_lef:
+        lef_args += " --input-lef {}".format(file.path)
+    gds_args = " "
+    for file in ctx.files.input_gds:
+        gds_args += " --input-gds {}".format(file.path)
+    gds_allow_empty_args = ""
+    if ctx.attr.gds_allow_empty:
+        gds_allow_empty_args = " --gds-allow-empty {}".format(ctx.attr.gds_allow_empty)
+
+    ctx.actions.run_shell(
+        outputs = [
+            final_gds
+        ],
+        inputs = depset(
+            ctx.files.input_lef +
+            ctx.files.input_gds +
+            [
+                ctx.attr.implemented_rtl[OpenRoadInfo].routed_def,
+                ctx.file.klayout_lyt,
+            ]
+        ),
+        command = "python {}".format(ctx.executable._gds_write.path) + \
+                  " --design-name {}".format(ctx.attr.implemented_rtl[SynthesisInfo].top_module) + \
+                  " --input-def {}".format(ctx.attr.implemented_rtl[OpenRoadInfo].routed_def.path) + \
+                  lef_args + \
+                  gds_args + \
+                  " --tech-file {}".format(ctx.file.klayout_lyt.path) + \
+                  gds_allow_empty_args + \
+                  " --out {}".format(final_gds.path),
+        tools = depset([ctx.executable._gds_write]),
+    )
+
+    return [DefaultInfo(files = depset([final_gds]))]
+
+gds_write = rule(
+    implementation = _gds_write_impl,
+    attrs = {
+        "implemented_rtl": attr.label(mandatory = True, providers = [OpenRoadInfo, SynthesisInfo]),
+        "klayout_lyt": attr.label(mandatory = True, allow_single_file = True),
+        "input_lef": attr.label_list(allow_files = True),
+        "input_gds": attr.label_list(allow_files = True),
+        "gds_allow_empty": attr.string(),
+        "_gds_write": attr.label(
```

Comment:
do we need this to be an attribute?

Reply:
This is used to create implicit dependency on def2stream py_binary. `_gds_write` is a private attribute so it is not possible to override its default value. This way we can also refer to the script through `ctx.executable.<_attr_name>`.

I saw a similar solution in [STA rule](https://github.com/hdl/bazel_rules_hdl/blob/main/static_timing/build_defs.bzl#L78-L82) and [Verilator rule](https://github.com/hdl/bazel_rules_hdl/blob/main/verilator/defs.bzl#L209-L214) and I wanted to follow this pattern here as well.

Reply:
LGTM.